### ISSUE TITLE
Always consider __del__ to be KeyboardInterrupt protected

### ISSUE THREAD

**Issue by alexshpilkin** [python-trio/trio#676]
I just tried to interrupt a long-running program of mine and got the following message:
```
^CException ignored in: <function Nursery.__del__ at 0x7fb575698510>
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/trio/_core/_run.py", line 520, in __del__
    def __del__(self):
  File "/usr/lib/python3.7/site-packages/trio/_core/_ki.py", line 191, in handler
    raise KeyboardInterrupt
```
The interrupt was indeed ignored. A second interrupt a dozen seconds later did the job; there were several more web requests issued from the main task in between. After reading the docs, I was under the impression that this was never, ever supposed to happen (unless done on purpose).

What puzzles me is that my program didn’t create any nurseries or Tasks _at all_. It did use `asks` with connections > 1, though, but did not—itself—issue more than one request at a time.

I’m sorry that I can’t reproduce this reliably, but my only evidence is a failure of a single several-hour-long run of a web scraper, with several previous runs coming out OK. I’m just writing this up in case it can be fixed by staring at the code for `Nursery.__del__`.

**Comment by njsmith** [comment]
Huh! That is an interesting corner case. Thanks for pointing this out.

So here are two general facts about python:

* `__del__` methods run in a strange context where they can execute arbitrary python code, but cannot raise exceptions

* Signal handlers, like the one that raises `KeyboardInterrupt`, can interrupt any python code

So i guess in any python program that uses `__del__`, there is some chance of `KeyboardInterrupt`s being lost, if you get really unlucky and hit control-C just as some `__del__` method runs. (Nothing special about `Nursery` here, it could be any python object with a `__del__`.) And trio does a lot of stuff to make control-C behave better, but since i never thought about this weird situation before, trio currently doesn't have any workaround for it, so it has the same flaw as regular python here. Neat! Also kind of annoying.

We could add a `@enable_ki_protection` decorator to all our `__del__` methods, but this still leaves a window where it can happen (because of `@enable_ki_protection`'s lack of atomicity), and also doesn't help with any other `__del__` methods in other libraries.

We could declare that all callables named `"__del__"` are automatically KI-protected. i guess this could go wrong in some edge cases:

* Someone calls `__del__` manually instead of through the GC: we can't tell who called it, so we have to assume it's the GC and defer raising the exception, even though in this case we could have safely raised

* Someone writes a `__del__` method that gets wedged and they are trying to kill it: normally control-C on a wedged `__del__` method won't exit the program, but it could at least force the `__del__` call to exit, so that the next control-C has a chance at working. If we make it KI-protected, then you can't even do this.

Neither of these seems to dire, though. Calling `__del__` directly is rare, and the consequences for deferring the raise are generally pretty minimal. And a wedged `__del__` is certainly possible in principle, but I've never seen it happen in real life, and there are plenty of more common ways to make it hard to kill a program.

So I guess adding an "are we in `__del__`?" check to the stack walking code in `_ki.py` would be good.

------

Also, while we're here: @alexshpilkin, thanks for using trio! I want to check on one thing that confused me a little: you're aware, right, that asks doesn't automatically parallelize your requests? It lets you run multiple requests simultaneously, but if you want to speed up your web crawl then you do still need to make a nursery and issue multiple requests at the same time.

**Comment by alexshpilkin** [comment]
@njsmith Sure (although that would be nice =) ). I was running a slow sequential traversal to make it a bit easier to see which pages will crash my parser. It’s awfully slow, though, so I guess I’ll still have to write some code to collect the error data from the parallel version.

Re: the strange context: Looks like there really is no way to detect it. The code that handles calling `__del__` (in [pyobject.c][1]) does not provide any indication that we can’t raise exceptions. (I was surprised by how many other occurrences of `PyErr_WriteUnraisable` there are in the code. I’ve no idea what they are doing.)

[1]: https://github.com/python/cpython/blob/b46ad5431d2643f61e929c1ffec48766b2fafd75/Objects/typeobject.c#L6660

**Comment by oremanj** [comment]
Renaming to reflect the apparent consensus (which I agree with).

**Comment by njsmith** [comment]
So the concrete steps here are:

* Modify the `current_ki_protected` function in `trio/_core/_ki.py` so that it returns `True` if it sees a function named `__del__`
* Add tests to make sure that this works. They go in `trio/_core/tests/test_ki.py`, and there are some examples to work from there. We'll want to test the cases: (1) a control-C arrives inside a `__del__` function, (2) a control-C arrives inside a function that was called from a `__del__` function.
* Add a [release note](https://trio.readthedocs.io/en/latest/contributing.html#release-notes)

**Comment by alexshpilkin** [comment]
@njsmith Erm.. Are you sure this works? As far as I can tell, you are talking about something like ...
```diff
--- a/trio/_core/_ki.py
+++ b/trio/_core/_ki.py
@@ -3,6 +3,7 @@ import signal
 import sys
 from contextlib import contextmanager
 from functools import wraps
+from inspect import getframeinfo
 
 import async_generator
 
@@ -90,6 +91,8 @@ LOCALS_KEY_KI_PROTECTION_ENABLED = '@TRIO_KI_PROTECTION_ENABLED'
 # this function:
 def ki_protection_enabled(frame):
     while frame is not None:
+        if getframeinfo(frame).function.__name__ = '__del__':
+            return True
         if LOCALS_KEY_KI_PROTECTION_ENABLED in frame.f_locals:
             return frame.f_locals[LOCALS_KEY_KI_PROTECTION_ENABLED]
         frame = frame.f_back
```
... but there’s no guarantee that a `__del__` class attribute actually refers to a function object with `__name__` set to `"__del__"`: if I do `Class.__del__ = lambda: ...`, it won’t.

It’s an unlikely problem, but it does appear to be one of correctness. I don’t see how to handwave it away, unlike delayed delivery (and that one’s inevitable).

(The patch above is actually wrong if `LOCALS_KEY_KI_PROTECTION_ENABLED` can be false—we need two separate loops then—but it’s 1 AM here, sorry.)

**Comment by njsmith** [comment]
@alexshpilkin It would be a heuristic, yes – if someone has a function that happens to be named `__del__`, but isn't being called from the GC, then we might delay delivering `KeyboardInterrupt` unnecessarily. I think this is ok though: generally we still deliver it pretty soon, and signals don't guarantee that they'll be processed at a certain exact millisecond. If someone writes an infinite loop, and puts it inside a `__del__` method, that isn't actually a `__del__` method, then that might be a problem... But this seems like it would be incredibly rare. And fundamentally the whole `KeyboardInterrupt` mechanism is a best effort thing anyway....

**Comment by alexshpilkin** [comment]
@njsmith I’m talking about the opposite thing, like so:

```python
class C(object):
        pass
def f(self):
        print('I am ' + C.__del__.__name__)  # prints "I am f"
        print('Will you interrupt me?')
C.__del__ = f

x = C()
del x
```

My patch above (for example) will not recognize this `__del__` as a `__del__`, and I’m not sure it’s possible to do so (without inspecting the bytecode).

**Comment by njsmith** [comment]
@alexshpilkin Yes, that's another corner case where this heuristic would fail.

**Comment by alexshpilkin** [comment]
@njsmith I’d think the real problem is that the interpreter lies about `__del__` running on the main stack—it should’ve used a separate stack, or maybe pushed a special frame. But getting this into CPython would probably be a hassle.

**Comment by pquentin** [comment]
As someone who is not familiar with these issues, I'm grateful that you pointed out those corner cases, @alexshpilkin. Being able to see those issues helps to be a great trio developer :)

But I also agree with what appears to be the main consensus: the heuristic would not work in all cases, but would still be helpful. We need to be pragmatic here!

Are you interested in putting together a pull request?

**Comment by njsmith** [comment]
Fixed in #1040
